kube/src/config/mod.rs

```diff
@@ -14,10 +14,60 @@ use crate::{Error, Result};
 pub use file_loader::KubeConfigOptions;
 use file_loader::{ConfigLoader, Der};
 
+use chrono::{DateTime, Utc};
 use reqwest::header::{self, HeaderMap};
+use tokio::sync::Mutex;
 
+use std::sync::Arc;
 use std::time::Duration;
 
+#[derive(Debug, Clone)]
+pub(crate) enum Authentication {
+    None,
+    Basic(String),
+    Token(String),
+    RefreshableToken(Arc<Mutex<(String, DateTime<Utc>)>>, ConfigLoader),
+}
+
+impl Authentication {
+    async fn to_header(&self) -> Result<Option<header::HeaderValue>> {
+        match self {
+            Self::None => Ok(None),
+            Self::Basic(value) => {
+                Ok(Some(header::HeaderValue::from_str(value).map_err(|e| {
+                    Error::Kubeconfig(format!("Invalid basic auth: {}", e))
+                })?))
+            }
+            Self::Token(value) => {
+                Ok(Some(header::HeaderValue::from_str(value).map_err(|e| {
+                    Error::Kubeconfig(format!("Invalid bearer token: {}", e))
+                })?))
+            }
+            Self::RefreshableToken(data, loader) => {
+                let mut locked_data = data.lock().await;
```

Comment:
So this is the lock that's being gated at via the `client` module. Presumably if many things happen at the same time, and we are awaiting a refresh token, then they all queue up here, and clash?

Reply:
They will queue up here but they shouldn't clash. They are all futures, so they'll just block unless the lock is released by the previous owner. I would use a `RwLock` here, but we have to check for expiry, so we'd have to handle an "upgrade" to a write lock in the case that we need a refresh

Reply:
As long as the queued up checks don't all re-trigger the same action this all makes sense :+1:

Reply:
The queued checks will only trigger the lock, it will not trigger a token refresh, so this should be ok